diesel/src/migrations/mod.rs

```diff
@@ -91,27 +93,128 @@ use std::path::{PathBuf, Path};
+/// any migration fails to run. Each migration is run in its own transaction, so some migrations
+/// may be committed, even if a later migration fails to run. If there was no Err a Vec with versioncodes 
+/// of executed migrations will be returned.
+/// 
+/// ```rust
+/// # extern crate diesel;
+/// # include!("src/doctest_setup.rs");
+/// #
+/// include!(concat!(env!("OUT_DIR"), "/migrations.rs"));
+/// 
+/// use self::migrations::run;
+/// let connection = establish_connection();
+/// run(&conn).unwrap();
+/// ```
+/// See the [module level documentation](index.html) for information on how migrations should be
+/// structured, and where Diesel will look for them by default.
+
+pub fn generate_migrations_module(dst: &Path) -> Result<(), RunMigrationsError> {
+    let migration_module=try!(build_mirgations_module());
+    let mut f = try!(File::create(dst));
+    try!(f.write_all(&migration_module.as_bytes()));
+    Ok(())
+}
+
+
 #[doc(hidden)]
-pub fn run_pending_migrations_in_directory<Conn>(conn: &Conn, migrations_dir: &Path, output: &mut Write)
-    -> Result<(), RunMigrationsError> where
+pub fn run_pending_migrations_form_iter<Conn, T, M>(conn: &Conn, all_migrations:T, output: &mut Write)-> Result<Vec<String>, RunMigrationsError> where
         Conn: MigrationConnection,
-{
+        T: Iterator<Item=M>,
+        M: Migration{
+
     try!(create_schema_migrations_table_if_needed(conn));
     let already_run = try!(conn.previously_run_migration_versions());
-    let all_migrations = try!(migrations_in_directory(migrations_dir));
-    let pending_migrations = all_migrations.into_iter().filter(|m| {
+    let pending_migrations = all_migrations.filter(|m| {
         !already_run.contains(&m.version().to_string())
     });
     run_migrations(conn, pending_migrations.collect(), output)
 }
 
+#[doc(hidden)]
+pub fn run_pending_migrations_in_directory<Conn>(conn: &Conn, migrations_dir: &Path, output: &mut Write)
+    -> Result<Vec<String>, RunMigrationsError> where
+        Conn: MigrationConnection,
+{
+    use self::migration::migration_from;
+
+    let all_migrations = try!(migrations_in_directory(migrations_dir,&migration_from));
+    run_pending_migrations_form_iter(conn,all_migrations.into_iter(),output)
+}
+
+fn build_mirgations_module() -> Result<String, MigrationError>{
+
+    let migrations_dir = try!(find_migrations_directory());
+    build_migrations_module_form_directory(&migrations_dir)
+}
+
+fn join<S:ToString> (l: Vec<S>, sep: &str) -> String{
+    l.iter().fold("".to_string(),
+                  |a,b| if a.len()>0 {a+sep} else {a} + &b.to_string()
+                  )
+}
+
+fn build_migrations_module_form_directory(migrations_dir: &Path) -> Result<String, MigrationError>
+{
+    use self::migration::programmatic_migration_from;
+
+    let migrations = try!(migrations_in_directory(migrations_dir,&programmatic_migration_from));
+    Ok(format!("pub mod migrations {{
```

Comment:
Same thing here. We shouldn't be creating code using string munging, this needs to go in codegen.